src/test/FreeC/Pass/ExportPassTests.hs

```diff
@@ -0,0 +1,126 @@
+                                                 (Qual "A" (Ident "Foo"))
+                                                 (fromJust mOutput)
+        entryIdent foo `shouldBeQualifiedWith` "A"
+    it "qualifies Coq identifiers of constructor entries" $ do
+      input <- expectParseTestModule ["module A where", "data A.Bar = A.Foo"]
+      shouldSucceedWith $ do
+        _       <- defineTestTypeCon "A.Bar" 0 ["A.Foo"]
+        _       <- defineTestCon "A.Foo" 0 "A.Bar"
+        _       <- exportPass input
+        mOutput <- inEnv $ lookupAvailableModule "A"
+        let foo = fromJust $ lookupExportedEntry IR.ValueScope
+                                                 (Qual "A" (Ident "Foo"))
+                                                 (fromJust mOutput)
+        entryIdent foo `shouldBeQualifiedWith` "A"
+    it "qualifies Coq identifiers of type synonym entries" $ do
+      input <- expectParseTestModule
+        ["module A where", "data A.Bar = A.Bar;", "type A.Foo = A.Bar"]
+      shouldSucceedWith $ do
+        _       <- defineTestTypeCon "A.Bar" 0 ["A.Bar"]
+        _       <- defineTestCon "A.Bar" 0 "A.Bar"
+        _       <- defineTestTypeSyn "A.Foo" [] "A.Bar"
+        _       <- exportPass input
+        mOutput <- inEnv $ lookupAvailableModule "A"
+        let foo = fromJust $ lookupExportedEntry IR.TypeScope
+                                                 (Qual "A" (Ident "Foo"))
+                                                 (fromJust mOutput)
+        entryIdent foo `shouldBeQualifiedWith` "A"
+    it "qualifies Coq identifiers of function declaration entries" $ do
+      input <- expectParseTestModule
+        ["module A where", "data A.Bar = A.Bar;", "type A.Foo = A.Bar"]
+      shouldSucceedWith $ do
+        _       <- defineTestTypeCon "A.Foo" 0 ["A.Foo"]
+        _       <- defineTestCon "A.Foo" 0 "A.Foo"
+        _       <- defineTestFunc "A.mkFoo" 0 "A.Foo"
+        _       <- exportPass input
+        mOutput <- inEnv $ lookupAvailableModule "A"
+        let mkFoo = fromJust $ lookupExportedEntry IR.ValueScope
+                                                   (Qual "A" (Ident "mkFoo"))
+                                                   (fromJust mOutput)
+        entryIdent mkFoo `shouldBeQualifiedWith` "A"
+    it "does not override qualification of Coq identifiers for entries" $ do
+      _     <- expectParseTestModule ["module A where", "data A.Foo = A.Foo"]
+      input <- expectParseTestModule ["module B where", "import A"]
+      shouldSucceedWith $ do
+        _    <- defineTestTypeCon "A.Foo" 0 ["A.Foo"]
+        _    <- defineTestCon "A.Foo" 0 "A.Foo"
+        name <- parseTestQName "A.Foo"
+        let qualifier = Coq.ident "A"
+            textName  = Coq.ident "Foo"
+        modifyEnv $ modifyEntryIdent IR.TypeScope
+                                     name
+                                     (Coq.Qualified qualifier textName)
```

Comment:
You could save a lot of vertical space by defining the entrire `Coq.Qualid` in the `let` expression.

```suggestion
        let qualid = Coq.Qualified (Coq.ident "A") (Coq.ident "Foo")
        modifyEnv $ modifyEntryIdent IR.TypeScope name qualid
```